internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+	idCert *x509.Certificate
+	// The identity key used for signing TLS certificates.
+	signer crypto.Signer
+}
+
+// NewManager create a new TLS keypair manager. The tlsDir must be a
+// directory of which only the leaf element, at most, does not exist. The
+// leaf directory will be created with 0o700 permissions. The signer
+// must implement a crypto.Signer interface. The signer represents the
+// identity key of the machine, container or device. The signer will be
+// used to sign TLS keypairs and its certificate (CA) is included in the
+// TLS certificate chain, allowing a client to pin the identity
+// certificate once trust has been manually established as part of a
+// trust exchange procedure.
+func NewManager(tlsDir string, signer crypto.Signer) (*TLSManager, error) {
+	manager := &TLSManager{
+		tlsDir: tlsDir,
+		signer: signer,
+	}
+	return manager, nil
+}
+
+// GetCertificate returns an identity signed TLS certificate. The certificate chain includes
+// both the TLS leaf certificate, as well as the self signed identity CA certificate. If
+// either the identity or TLS certificate nears expiry, this functions creates new
+// certificates on demand. Note that even if the identity certificate is re-created, this
+// does not mean that the identity key changed (the key itself has no expiry).
+func (m *TLSManager) GetCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	var err error
+
+	// Fast path: concurrent sessions while the ID and TLS certificate is valid.
+	m.mu.RLock()
+	tlsCert := m.tlsCert
+	idCert := m.idCert
+	m.mu.RUnlock()
+	if idCert != nil && isCertActive(idCert) && tlsCert != nil && isCertActive(tlsCert.Leaf) {
+		return tlsCert, nil
+	}
+
+	// Slow path: generate a new in-memory identity signed TLS certificate.
+	//
+	// If we got here then it means we need to generate a new in-memory TLS
+	// keypair, and potentially an identity certificate (only the first time
+	// or when it expires, or when the key changed).
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if err := m.createDir(); err != nil {
+		return nil, fmt.Errorf("cannot create TLS directory: %w", err)
+	}
+	if err := m.getIDCert(); err != nil {
+		return nil, fmt.Errorf("cannot get identity certificate: %w", err)
+	}
+	if err = m.createTLSCert(); err != nil {
+		return nil, fmt.Errorf("cannot create TLS certificate: %w", err)
+	}
+	return m.tlsCert, nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *TLSManager) Ensure() error {
+	return nil
+}
+
+func (m *TLSManager) createTLSCert() error {
+	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return err
+	}
+	template := x509.Certificate{
+		SerialNumber:          serialNumber,
+		NotBefore:             systemTime(),
+		NotAfter:              systemTime().Add(tlsCertValidity),
```

Comment:
For sanity, it's probably best to call `now := systemTime()` once, and then use that one value for both of these, rather than have two values a little bit out.

Reply:
Yes. I fixed this in two places, thanks. Resolving.